nodejs - Optimize copy for buildah

neeraj-laad opened the pull request:
We are seeing significant performance issues with all Node.js based stacks to build the final application image when using buildah (instead of docker).

This is because of a known issue with buildah 1.9.0. We can not upgrade the buildah version because of other issues. 

This is a temporary fix to optimize the number of files we copy during the build process.

### Checklist:

- [x] Read the [Code of Conduct](https://github.com/appsody/website/blob/master/CODE_OF_CONDUCT.md) and [Contributing Guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md).

- [x] Followed the [commit message guidelines](https://github.com/appsody/website/blob/master/CONTRIBUTING.md#commit-message-guidelines).

- [x] Stack adheres to [Appsody stack structure](https://github.com/appsody/website/blob/master/content/docs/stacks/stacks-overview.md#stack-structure).

### Modifying an existing stack:

- [x] Updated the stack version in `stack.yaml`

### Related Issues:
<!-- e.g. Fixes #32, Related to #54, etc. -->

sam-github commented:
This PR is a duplicate of https://github.com/appsody/stacks/pull/680 which changes both stacks in the same PR. I suggest closing this PR, since the changes are cut-n-paste, its easier to fix in one PR and easier to understand the historical record as well.

neeraj-laad commented:
@sam-github I have re-worked this PR, as we were copying `node_modules` from user-app over in the final stage of the build. This meant that dev/test dependencies from the user's development environment were getting into the final application image. 

I have now moved the copying of the application over in the first stage and I delete `node_modules` before running `npm i --production`. This has also cleaned up the copy steps a bit. 

Please let me know your thoughts.

sam-github commented on incubator/nodejs/image/project/Dockerfile:
```diff
@@ -7,11 +7,18 @@ RUN apt-get update \
  && echo 'Finished installing dependencies'
 
+# Copying individual files/folders as buildah 1.9.0 does not honour .dockerignore
+COPY --chown=node:node package*.json /project/
```
These package files don't seem to be used (there is no `WORKDIR /project; RUN npm install`). The old code used to do `COPY --chown=node:node . /project`, so it would be copying `COPY package.* node_modules /project`, but that doesn't seem to be replaced. Perhaps that happens elsewhere? Sorry, I only have a general sense of how appsody uses this file.

neeraj-laad replied:
The intent is to copy the contents of /project and /project/user-app but not copy any `/node_modules`. 

Given /project contains files from the stack, I can achieve this by copying selective files. User's appl could be in any structure so I'm copying the entire user-app folder and then delete the user-app/node_modules folder later.

That said I agree that we are not using `/project/package.json` at all. It is only used during development and adds appmetrics-dash. So we can skip copying it here.

neeraj-laad commented:
@sam-github Would like to get a final review from you on this.
